src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,472 @@
+      auto geom = domain.GetFaceGeometricFactors(ir, flags, mfem::FaceType::Boundary);
+      boundary_integrals_.emplace_back(num_boundary_elements, geom->J, geom->X, Dimension<geometry_dim>{},
+                                       Dimension<spatial_dim>{}, integrand);
+      return;
+    } else if constexpr (true) {
+      // if static_assert has a literal 'false' for its first arg,
+      // it will trigger even when this branch isn't selected. So,
+      // we define an expression that is always false (see first
+      // if constexpr expression) to work around this limitation
+      constexpr bool always_false = (geometry_dim == spatial_dim);
+      static_assert(always_false, "unsupported integral dimensionality");
+    }
+  }
+
+  /**
+   * @brief Adds an area integral, i.e., over 2D elements in R^2 space
+   * @param[in] integrand The quadrature function
+   * @param[in] domain The mesh to evaluate the integral on
+   */
+  template <typename lambda>
+  void AddAreaIntegral(lambda&& integrand, mfem::Mesh& domain)
+  {
+    AddIntegral(Dimension<2>{} /* geometry */, Dimension<2>{} /* spatial */, integrand, domain);
+  }
+
+  /**
+   * @brief Adds a volume integral, i.e., over 3D elements in R^3 space
+   * @param[in] integrand The quadrature function
+   * @param[in] domain The mesh to evaluate the integral on
+   */
+  template <typename lambda>
+  void AddVolumeIntegral(lambda&& integrand, mfem::Mesh& domain)
+  {
+    AddIntegral(Dimension<3>{} /* geometry */, Dimension<3>{} /* spatial */, integrand, domain);
+  }
+
+  /**
+   * @brief Adds a domain integral
+   * @tparam d The dimension of the elements *and* the space they're embedded in
+   * @param[in] integrand The quadrature function
+   * @param[in] domain The mesh to evaluate the integral on
+   */
+  template <int d, typename lambda>
+  void AddDomainIntegral(Dimension<d>, lambda&& integrand, mfem::Mesh& domain)
+  {
+    AddIntegral(Dimension<d>{} /* geometry */, Dimension<d>{} /* spatial */, integrand, domain);
+  }
+
+  /**
+   * @brief Adds a surface integral, i.e., over 2D elements in R^3 space
+   * @param[in] integrand The quadrature function
+   * @param[in] domain The mesh to evaluate the integral on
+   */
+  template <typename lambda>
+  void AddSurfaceIntegral(lambda&& integrand, mfem::Mesh& domain)
+  {
+    AddIntegral(Dimension<2>{} /* geometry */, Dimension<3>{} /* spatial */, integrand, domain);
+  }
+
+  /**
+   * @brief Implements mfem::Operator::Mult
+   * @param[in] input_T The input vector
+   * @param[out] output_T The output vector
+   */
+  void Mult(const mfem::Vector& input_T, mfem::Vector& output_T) const override
+  {
+    Evaluation<Operation::Mult>(input_T, output_T);
+  }
+
+  /**
+   * @brief Implements mfem::Operator::GetGradient
+   * @param[in] x The input vector
+   */
+  mfem::Operator& GetGradient(const mfem::Vector& x) const override
+  {
+    Mult(x, dummy_);  // this is ugly
+    return grad_;
```

Comment:
I'm a bit confused by this. `grad_` appears to be a reference to this object, but `Mult()` calls down to `Operation::Mult`. Some clarifying comments would be appreciated.